### PR TITLE
Ad email validation

### PR DESCRIPTION
**Pull request by apalski**
PR to add email confirmation and password length validations to user model:
- Add email_confirmation to user_params in admin/user controller and user controller
- Add validations to user model
- Add email_confirmation field to admin/user and user views
- Add password to user edit view
- Add error messages to locales/en.yml
- Add email_comfirmation to admin/user and user feature specs
- Add password to user edit feature spec
- Add validation specs to user model spec
https://trello.com/c/kWq6ngnN
https://trello.com/c/3y1ZJkqg
All feedback and suggestions welcome - thanks for taking a look!

### CODE REVIEW

**Comment by scara82** [comment]
LGTM! 👍

File: app/controllers/admin/users_controller.rb [hunk 1]
```diff
@@ -36,7 +36,14 @@ def destroy
...
 
     def user_params
-      params.require(:user).permit(:name, :email, :password, :uid, :admin)
+      params.require(:user).permit(
+        :name,
+        :email,
+        :password,
+        :uid,
+        :admin,
+        :email_confirmation
```
Comment (elle) [inline app/controllers/admin/users_controller.rb]: comma.
also alphabetical order?

File: app/controllers/users_controller.rb [hunk 2]
```diff
@@ -7,6 +7,9 @@ def new
 
   def create
     @user = User.create(user_params)
+    if @user
+      session[:user_id] = @user.id
+    end
```
Comment (elle) [inline app/controllers/users_controller.rb]: This does not belong in this PR (if this PR is about email validation).

Also, since this is logging the user in, then a suggestion to pull out this into its own method that can be called here and in the sessions controller.

File: app/controllers/users_controller.rb [hunk 3]
```diff
@@ -26,6 +29,12 @@ def update
   private
 
   def user_params
-    params.require(:user).permit(:name, :email, :password, :uid)
+    params.require(:user).permit(
+      :name,
+      :email,
+      :password,
+      :uid,
+      :email_confirmation
```
Comment (elle) [inline app/controllers/users_controller.rb]: comma?
Alphabetical order?

File: app/models/user.rb [hunk 4]
```diff
@@ -4,6 +4,8 @@ class User < ActiveRecord::Base
   has_one :budget
 
   validates :name, :email, presence: true, uniqueness: true
+  validates :email, confirmation: true
+  validates :password, length: { in: 6..20 }
```
Comment (elle) [inline app/models/user.rb]: Does password validation belong in this PR?

File: spec/features/users/user_updates_profile_spec.rb [hunk 5]
```diff
@@ -7,11 +7,12 @@
 
     visit edit_users_path
 
-    fill_in "user_email", with: "JT@example.com"
-    fill_in "user_name", with: user.name
+    fill_in "user_name", with: "Fred"
+    fill_in "user_email", with: user.email
+    fill_in "user_password", with: user.password
```
Comment (elle) [inline spec/features/users/user_updates_profile_spec.rb]: Do these two fields need to be filled manually? can the form work without them?

Websites usually have a different form/workflow for password changing. Maybe think about what the user experience you wish to create.

And a side note: implement reseting your password, if the user cannot log in (in a future PR)

**Comment by elle** [comment]
The name of the PR is confusing. This PR is about email confirmation, not email validation :)

File: spec/models/user_spec.rb [hunk 6]
```diff
@@ -8,6 +8,8 @@
     it { is_expected.to validate_presence_of(:email) }
     it { is_expected.to validate_uniqueness_of(:name) }
     it { is_expected.to validate_uniqueness_of(:email) }
+    it { is_expected.to validate_confirmation_of(:email) }
+    it { is_expected.to validate_length_of(:password) }
```
Comment (elle) [inline spec/models/user_spec.rb]: A suggestion to add: `.is_at_least(6).is_at_most(20)`

File: app/models/user.rb (hunk 4 above)
Comment (elle) [reply app/models/user.rb]: If limiting password length, I would limit it at a higher number, for example 64. 20 is not secure enough.

You might want to add other validation rules to password, for example using letters and numbers. Maybe in another PR?

File: spec/features/users/user_creates_new_account_spec.rb [hunk 7]
```diff
@@ -6,14 +6,12 @@
...
     fill_in "user_password", with: "secret"
     click_on I18n.t("helpers.submit.user.create")
 
-    expect(page).to have_text(I18n.t("sessions.new.title"))
+    expect(page).to have_text(I18n.t("users.show.welcome"))
     expect(page).
-      to have_text(I18n.t(
-        "flash.actions.create.notice",
-        resource_name: "User"
-      ))
```
Comment (elle) [inline spec/features/users/user_creates_new_account_spec.rb]: Now that we are signing a user in after creation, do we want to check that this flash message was not on the page?

File: spec/features/users/user_creates_new_account_spec.rb [hunk 8]
```diff
@@ -6,14 +6,12 @@
...
+      to have_text(I18n.t("sessions.new.flash.alert"))
```
Comment (elle) [inline spec/features/users/user_creates_new_account_spec.rb]: Do we also want to check that the `current_path` is the user profile page?
i.e. `(expect(current_path).to eq user_path`

File: app/controllers/admin/users_controller.rb (hunk 1 above)
Comment (apalski) [reply app/controllers/admin/users_controller.rb]: Done in sha: 1159cb8

File: app/controllers/users_controller.rb (hunk 2 above)
Comment (apalski) [reply app/controllers/users_controller.rb]: Sorry, didn't really name this PR well at all, it was meant to be a general validation/confirmation PR.
Done in sha: 1159cb8

File: app/controllers/users_controller.rb (hunk 3 above)
Comment (apalski) [reply app/controllers/users_controller.rb]: Done in sha: 1159cb8

File: app/models/user.rb (hunk 4 above)
Comment (apalski) [reply app/models/user.rb]: Done in sha: 1159cb8
Will look at another PR to set stronger validation rules for password.

File: spec/features/users/user_updates_profile_spec.rb (hunk 5 above)
Comment (apalski) [reply spec/features/users/user_updates_profile_spec.rb]: The email one didn't, it is populated when you arrive at the form but the password isn't?
I will look at password reset and changing in a later PR.
Done in sha: 1159cb8

File: spec/models/user_spec.rb (hunk 6 above)
Comment (apalski) [reply spec/models/user_spec.rb]: Done in sha: 1159cb8

File: spec/features/users/user_creates_new_account_spec.rb (hunk 7 above)
Comment (apalski) [reply spec/features/users/user_creates_new_account_spec.rb]: Done in sha: 1159cb8

File: spec/features/users/user_creates_new_account_spec.rb (hunk 8 above)
Comment (apalski) [reply spec/features/users/user_creates_new_account_spec.rb]: Done in sha: 1159cb8

File: config/locales/en.yml [hunk 9]
```diff
@@ -109,7 +109,10 @@ en:
     errors:
       messages:
         blank: "can't be blank"
+        confirmation: "email must match"
         taken: "already taken"
+        too_short: "minimum length 6"
+        too_long: "maximum length 20"
```
Comment (elle) [inline config/locales/en.yml]: Should be updated to 64 :)

File: spec/features/users/user_updates_profile_spec.rb [hunk 10]
```diff
@@ -7,11 +7,11 @@
 
     visit edit_users_path
 
-    fill_in "user_email", with: "JT@example.com"
-    fill_in "user_name", with: user.name
+    fill_in "user_name", with: "Fred"
+    fill_in "user_password", with: user.password
```
Comment (elle) [inline spec/features/users/user_updates_profile_spec.rb]: I know I mentioned this before. Adding this field here, means a user cannot update their profile without inputing in their password. Side note: suggesting to fix in the future.

File: config/locales/en.yml (hunk 9 above)
Comment (apalski) [reply config/locales/en.yml]: Phew, thanks for picking that up @elle!

File: spec/features/users/user_updates_profile_spec.rb (hunk 10 above)
Comment (apalski) [reply spec/features/users/user_updates_profile_spec.rb]: Sorry, slow on the uptake, will fix in another PR!

**Comment by apalski** [comment]
Merging now, thank you for the feedback @elle and @scara82 !
